business/authentication/openid_auth_controller.go

```diff
@@ -0,0 +1,1444 @@
+			return fmt.Errorf("we don't like people living in the future - enjoy the present!; iat = '%d'", parsedIat)
+		}
+	}
+
+	// Check exp (expiration time) claim
+	// The OIDC spec says: "The current time MUST be before the time represented by the exp Claim"
+	// No tolerance for this check.
+	if !util.Clock.Now().Before(openIdParams.ExpiresOn) {
+		return fmt.Errorf("the OpenId token has expired; exp = '%s'", openIdParams.ExpiresOn.String())
+	}
+
+	// There are other claims that could be checked, but are not verified here:
+	//   - nonce: This should be verified regardless if RBAC is on/off. So, it's verified in
+	//       another part of the authentication flow.
+	//   - acr: we are not asking for this claim at authorization, so the IdP doesn't
+	//       need to provide it nor we need to verify it.
+	//   - auth_time: we are not asking for this claim at authorization, so the IdP doesn't
+	//	     need to provide it nor we need to verify it.
+
+	// If execution flow reached this point, all claims look valid, but that won't guarantee that
+	// the id_token hasn't been tampered. So, we check the signature to find if
+	// the token is genuine
+	if kidHeader := openIdParams.ParsedIdToken.Headers[0].KeyID; len(kidHeader) == 0 {
+		return errors.New("the OpenId token is missing the kid header claim")
+	} else {
+		if jws, parseErr := jose.ParseSigned(openIdParams.IdToken); parseErr != nil {
+			return fmt.Errorf("error when parsing the OpenId token: %w", parseErr)
+		} else {
+			if len(jws.Signatures) == 0 {
+				return errors.New("an unsigned OpenId token is not acceptable")
+			}
+
+			matchingKey, findKeyErr := getJwkFromKeySet(kidHeader)
+			if findKeyErr != nil {
+				return fmt.Errorf("something went wrong when trying to find the key that signed the OpenId token: %w", findKeyErr)
+			}
+			if matchingKey == nil {
+				return errors.New("the OpenId token is signed with an unknown key")
+			}
+
+			_, signVerifyErr := jws.Verify(matchingKey)
+			if signVerifyErr != nil {
+				return fmt.Errorf("the signature of the OpenId token is invalid: %w", signVerifyErr)
+			}
+		}
+	}
+
+	return nil
+}
+
+// verifyOpenIdUserAccess checks that the provided token has enough privileges on the cluster to
+// allow a login to Kiali.
+func verifyOpenIdUserAccess(token string, businessInstantiator func(authInfo *api.AuthInfo) (*business.Layer, error)) (int, string, error) {
+	// Create business layer using the id_token
+	bsLayer, err := businessInstantiator(&api.AuthInfo{Token: token})
+	if err != nil {
+		return http.StatusInternalServerError, "Error instantiating the business layer", err
+	}
+
+	// Using the namespaces API to check if token is valid. In Kubernetes, the version API seems to allow
+	// anonymous access, so it's not feasible to use the version API for token verification.
+	nsList, err := bsLayer.Namespace.GetNamespaces(context.TODO())
+	if err != nil {
+		return http.StatusUnauthorized, "Token is not valid or is expired", err
+	}
```

Comment:
Do we need to also check other services i.e. prom, jaeger?

Reply:
I don't think it's needed, because those services does not really support RBAC. Kiali, mostly uses the Kubernetes API to decide what to show/hide. If we were to only use Prometheus, it would happily return any requested data.